File: Week2/readability/readability.c
#include <cs50.h>
#include <ctype.h>
#include <math.h>
#include <stdio.h>
#include <string.h>

float compute_index(string text);

int compute_letters(string text);
int compute_words(string text);
int compute_sentence(string text);

float compute_l(int words, int letters);
float compute_s(int words, int sentences);

void print_grade(float index);

int main(void)
{
    // 输入
    string text = get_string("Text: ");

    // 衡量指标
    float index = compute_index(text);

    // 输出等级
    print_grade(index);
}

float compute_index(string text)
{
    // 根据空格, 把字符数组, 每 100 词分为一组, 100 words, 100 spaces
    // 遍历一遍记录索引, 相比每 100 词存为新字符串更快
    // int per_hundred_words_index[] =
    // int split_index = split_text(text);
    // 上面思路没必要,缺点是每 100words 的计算结果可能不同，还得取平均

    // 可以计算出总 words 数 / 100, 得出相对应 100 words 有多少 letters 和 sentens

    // 分别计算 letters, words, sentence
    int letters = compute_letters(text);
    int words = compute_words(text);
    int sentence = compute_sentence(text);

    // 计算 L
    float l = compute_l(words, letters);
    // 计算 S
    float s = compute_s(words, sentence);

    // 计算 index
    return (0.0588 * l - 0.296 * s - 15.8);
}

// int split_text(string text)
// {
//     // 记录每 100 words 的索引数组
//     int per_houndred_words_index[];
//     for (int i = 0, int count = 0, len = strlen(text); i < len; i++)
//     {
//         if (text[i] == ' ')
//         {
//             count++;
//         }
//         if ( count % 100 == 0)
//         {
//             per_houndred_words_index[] = i;
//         }
//     }
//     return
// }

int compute_letters(string text)
{
    int count = 0;
    for (int i = 0, len = strlen(text); i < len; i++)
    {
        if (tolower(text[i]) >= 'a' && tolower(text[i]) <= 'z')
        {
            count++;
        }
    }
    return count;
}

int compute_words(string text)
{
    int count = 1;
    for (int i = 0, len = strlen(text); i < len; i++)
    {
        if (text[i] == ' ')
        {
            count++;
        }
    }
    return count;
}

int compute_sentence(string text)
{
    int count = 0;
    for (int i = 0, len = strlen(text); i < len; i++)
    {
        if (text[i] == '.' || text[i] == '!' || text[i] == '?')
        {
            count++;
        }
    }
    return count;
}

float compute_l(int words, int letters)
{
    return letters / (words / 100.0);
}

float compute_s(int words, int sentences)
{
    return sentences / (words / 100.0);
}

void print_grade(float index)
{
    // 分三类, 两类特殊情况单独处理
    if (index <= 1)
    {
        printf("Before Grade 1\n");
    }
    else if (index >= 16)
    {
        printf("Grade 16+\n");
    }
    else
    {
        printf("Grade %i\n", (int) round(index));
    }
}

File: Week6/sentimental-readability/readability.py
from cs50 import get_string
import re


def cmpt_letter(text):
    n = 0
    for i in text:
        if i.isalpha():
            n += 1
    return n


def cmpt_word(text):
    word_list = text.split()    # 默认空格分隔句子为列表
    # print(word_list)
    return len(word_list)


def cmpt_sentence(text):
    sentence_list = re.split(r'[.\!?]', text)   # 以. ! ? 分割文本
    # print(sentence_list)
    return len(sentence_list) - 1               # 减去最后一个标点分割为俩


def cmpt_index(text):
    letters = cmpt_letter(text)
    words = cmpt_word(text)
    sentence = cmpt_sentence(text)

    # print("letters: {0}, words: {1}, sentence: {2}".format(letters, words, sentence))

    L = letters / (words / 100)
    S = sentence / (words / 100)

    # print("L: {0}, S: {1}".format(L, S))

    return 0.0588 * L - 0.296 * S - 15.8


def main():
    while True:
        text = get_string("Text: ")
        break

    index = round(cmpt_index(text))

    if index < 1:
        print("Before Grade 1")
    elif index >= 16:
        print("Grade 16+")
    else:
        print("Grade {0}".format(index))


if __name__ == "__main__":
    main()
